### ISSUE TITLE
[feature] implement TeaCache

### ISSUE THREAD

**Issue by sayakpaul** [huggingface/diffusers#12589]
Currently, we support PAB and FasterCache. More details: https://huggingface.co/docs/diffusers/main/en/optimization/cache. 

It'd be cool to support [Timestep Embedding Tells: It's Time to Cache for Video Diffusion Model](https://liewfeng.github.io/TeaCache/) as it's quite popular in the community. Code is here:
https://github.com/ali-vilab/TeaCache

Before proceeding with a PR, please propose a design first in this thread. We can take it from there.

**Comment by LawJarp-A** [comment]
Hey @sayakpaul 
I want to take this up, i have a few questions tho
- any preferred tests/datasets to test this on?
- it's model-specific due to different timestep embedding patterns across architectures. should I proto with CogVideoX first?

**Comment by LawJarp-A** [comment]
Additionally this is my thought process (I want to refine it still) @sayakpaul 

**Instead of monkey-patching model forward methods (like original TeaCache), use hooks:**
-> Create a `TeaCacheHook` that uses new_forward() to replace entire transformer block forward methods. The hook intercepts the block before any computation starts, extracts timestep embeddings, makes the skip/compute decision, and either returns cached residuals or runs the full computation.
-> Each supported model gets an extractor class that knows how to pull timestep embeddings from that model's specific architecture. CogVideoX extracts from `time_embedding(),` FLUX from `time_text_embed(),`

**State management**
-> Use diffusers existing BaseState system to track accumulated distance, previous embeddings, and cached residuals across inference steps (need to refine this a bit)

Add TeaCacheConfig to the existing cache configuration system in CacheMixin. This ensures it works seamlessly

Note: I am exploring the repository more as I work on figuring out how best to integrate this. I took some of time to test the current caching play around with it. I think this can be integrated well. I'm open to any suggestions and corrections

Thanks!

**Comment by yiyixuxu** [comment]
hi @LawJarp-A 
Thanks for your interest! 
feel free to open a PR and we can start working on it from there

**Comment by sayakpaul** [comment]
Yes, we can prototype with a model of choice. In case the original repository already provides results for a fairly popular model like Flux, let's prototype with that instead. If not, we can start with CogVideoX, as you mentioned.

**Comment by LawJarp-A** [comment]
Awesome, will prototype on flux first then. Will keep you guys updated on the PR
@sayakpaul @yiyixuxu 
Thanks :)

**Comment by LawJarp-A** [comment]
Working on this PR: https://github.com/huggingface/diffusers/pull/12652
As of now, writing it for just one model (FLUX), after testing for it, I want to write it with better strcuture to suuport future additions, something like below
```
  src/diffusers/hooks/teacache/
  ├── __init__.py              
  ├── base.py                  # TeaCacheConfig, TeaCacheHook base classes
  ├── extractors.py            # All model-specific extractors
  ├── flux.py                  
  ├── cogvideox.py         
and so on...
```

**Comment by sayakpaul** [comment]
Hey, 

Thanks for your willingness to contribute. Could we try to follow the hooks-based implementation for this (we use it for FasterCache)?

**Comment by LawJarp-A** [comment]
I am using hook based only now. Is there any issue in the PR?
(Not completed yet, still a draft PR)
(Above structure was something I was thinking out loud for the future expansion)

**Comment by sayakpaul** [comment]
Cool, we will check the PR first.

**Comment by AtulJoshi1206** [comment]
I'd like to help get TeaCache over the line.

@LawJarp-A already has this working in #12652 across FLUX, Mochi, Lumina2 and CogVideoX, with before/after images for all four — that's a lot of finished work and I don't want to duplicate it or talk over it. From reading the thread it doesn't look blocked on effort. It looks blocked on one open question: @sayakpaul flagged that the `.item()` on the skip decision makes the branch data-dependent under torch.compile, and the last thing on the thread is @LawJarp-A asking whether to go the `torch.cond` route — predicate stays a 0-dim bool, both branches traced, state updates gated with `torch.where` — or to do what `first_block_cache.py:133` and `mag_cache.py:184`/`:300` already do: put `@torch.compiler.disable` on the decision and keep the `.item()`. Both of those are merged and both make the same data-dependent branch, so I'm not sure whether TeaCache is being held to a higher bar deliberately or whether the question just got lost.

@LawJarp-A — are you still planning to take it forward? If you are, I'm happy to help with the compile refactor and the rebase; the branch conflicts with main now. If you'd rather hand it off, I'll pick it up from your branch with your commits intact.

So: does TeaCache need the `torch.cond` version, or is the `@torch.compiler.disable` approach that first_block_cache and mag_cache already use fine for a first merge?

**Comment by LawJarp-A** [comment]
yes, still on it. not handing it off, but help is welcome :)

one precision on the precedent: first_block_cache.py:133 disables only _should_compute_remaining_blocks, so the block forwards around it still compile. mag_cache.py:184/:300 disables the whole new_forward, which is a much bigger carve-out. teacache's decision site is shaped like the fbc one, so that's the version i'd ship.

my position: torch.cond is the right end state but shouldn't gate a first merge. it means pulling StateManager mutation out of both branches and threading it as carry-in/carry-out across four per-model forwards, for a property none of the merged cache hooks hold today. easier to review as its own pr with compile benchmarks than folded into this one.

@sayakpaul so: is the scoped @torch.compiler.disable (fbc-style, decision helper only) fine for merge, with torch.cond as a follow-up across all three cache hooks? or do you want teacache held to the torch.cond bar now? happy either way, lemme know!

@AtulJoshi1206 i'll take the rebase since it's my branch. if you want a piece, a torch.compile benchmark of the current branch on flux (compile hit + steady-state step time, cache on/off) would be the most useful thing, that's the number this hinges on. will ping here once the rebase is up.
